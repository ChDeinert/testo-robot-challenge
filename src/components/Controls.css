.controls__form {
  display: flex;
  flex-direction: column;
  gap: .5rem;
}

.controls__form__group {
  position: relative;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: baseline;
  gap: 1rem;
}

.controls__form__group__warning {
  position: absolute;
  right: 0;
  transform: translateX(calc(100% + 1rem));
  line-height: 1rem;
  padding: .25rem .5rem;
  border: 2px solid orangered;
  border-radius: 2px;
}

.controls__form__group__warning:empty {
  display: none;
}

.controls__form__label {
  font-weight: 600;
}

.controls__form__input {
  width: 5rem;
  padding: .25rem;
  background-color: white;
  border: 2px solid var(--color-blue);
  border-radius: .25rem;
}

.controls__form__input:focus {
  border: 2px solid var(--color-blue);
  outline: var(--color-orange) solid 2px;
  outline-offset: 2px;
}

.controls__form__input:invalid,
.controls__form__input--warning {
  box-shadow: 0 0 .25rem .125rem orangered;
}

.controls__form__input[type="submit"] {
  cursor: pointer;
  width: 100%;
  padding: .5rem;
  font-weight: 600;
}

.controls__form__input[type="submit"]:hover, 
.controls__form__input[type="submit"]:active {
  background-color: var(--color-blue-light);
}

.controls__buttons {
  display: flex;
  gap: .5rem;
}