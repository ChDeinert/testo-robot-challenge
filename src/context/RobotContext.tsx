import React, { Component, createContext, useContext } from 'react';
import Robot, { defaultBoard, possibleDirections, DirectionType, BoardType, PositionType } from '../lib/robot';

type RobotContextType = { 
  settings: {
    board: BoardType,
    possibleDirections: DirectionType[],
  },
  state: {
    initialized: boolean,
    isPlaced: boolean,
    position: { xPosition: number, yPosition: number, direction: DirectionType },
    error?: Error,
  },
  controls?: {
    place: (xCoordinate: number, yCoordinate: number, direction: DirectionType) => void,
    move: () => void,
    turnLeft: () => void,
    turnRight: () => void,
    report: () => undefined|PositionType,
  },
};

const defaultRobotContext : RobotContextType = {
  settings: {
    board: defaultBoard,
    possibleDirections: possibleDirections,
  },
  state: {
    initialized: false,
    isPlaced: false,
    position: { xPosition: 0, yPosition: 0, direction: possibleDirections[0] },
  },
};
const RobotContext = createContext<RobotContextType>(defaultRobotContext);
const useRobotContext = () => useContext(RobotContext);

/**
 * React component rendering the RobotContext provider,
 * it's setting, state and control-methods
 */
class RobotProvider extends Component {
  private robot?: Robot;
  state = {
    initialized: false,
    position: { xPosition: 0, yPosition: 0, direction: possibleDirections[0] },
    isPlaced: false,
    error: undefined
  };

  componentDidMount() {
    // Initialize the Robot only after the component is mounted
    // so that ssr/prerender aren't effected by any instance of Robot
    this.robot = new Robot(defaultBoard, possibleDirections);
    this.setState({ initialized: true });
  }

  updateCurrentPosition = () => {
    this.setState({ position: this.robot?.report(), error: undefined });
  }

  place = (xCoordinate: number, yCoordinate: number, facing: DirectionType) => {
    try {
      this.robot?.place(xCoordinate, yCoordinate, facing);
      this.updateCurrentPosition();
      this.setState({ isPlaced: true });
    } catch (error) {
      this.setState({ error });
    }
  }

  move = () => {
    try {
      this.robot?.move();
      this.updateCurrentPosition();
    } catch (error) {
      this.setState({ error });
    }
  };

  turnLeft = () => {
    try {
      this.robot?.left();
      this.updateCurrentPosition();
    } catch (error) {
      this.setState({ error });
    }
  };

  turnRight = () => {
    try {
      this.robot?.right();
      this.updateCurrentPosition();
    } catch (error) {
      this.setState({ error });
    }
  };

  report = () => {
    try {
      const currentPositionReport = this.robot?.report();
      alert(`The Robot's current position: ${currentPositionReport}`);
      return currentPositionReport;
    } catch (error) {
      this.setState({ error });
    }
  }

  render() { 
    const { children } = this.props;
    const { initialized, position, isPlaced, error } = this.state;

    return (
      <RobotContext.Provider value={{
        settings: { board: defaultBoard, possibleDirections: possibleDirections },
        state: {
          initialized,
          isPlaced,
          position,
          error,
        },
        controls: {
          place: this.place,
          move: this.move,
          turnLeft: this.turnLeft,
          turnRight: this.turnRight,
          report: this.report,
        }
      }}>
        {children}
      </RobotContext.Provider>
    );
  }
}

export default RobotContext;
export { RobotProvider, useRobotContext };