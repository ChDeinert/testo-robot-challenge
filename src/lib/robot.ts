type BoardType = { xWidth: number, yWidth: number };
type DirectionType = 'NORTH' | 'EAST' | 'SOUTH' | 'WEST';
type PositionType = { xPosition?: number, yPosition?: number, direction?: DirectionType, toString?: () => string };

const defaultBoard : BoardType = {
  xWidth: 5,
  yWidth: 5,
};

const possibleDirections = [
  'NORTH' as DirectionType, 'EAST' as DirectionType, 'SOUTH' as DirectionType, 'WEST' as DirectionType
];

const isRobotPlaced = (board : BoardType, possibleDirections: DirectionType[]) => (xCoordinate?: number, yCoordinate?: number, direction?: DirectionType) : boolean => {
  if (typeof xCoordinate !== 'number' || !isValidXCoordinate(xCoordinate, board) || 
      typeof yCoordinate !== 'number' || !isValidYCoordinate(yCoordinate, board) || 
      typeof direction === 'string' && !isValidDirection(direction, possibleDirections)) {
    return false;
  }

  return true;
};

const isValidXCoordinate = (xCoordinate: number, board: BoardType) : boolean => {
  if (typeof xCoordinate !== 'number' || isNaN(xCoordinate) || xCoordinate < 0 || xCoordinate >= board.xWidth) {
    return false;
  }

  return true;
};

const isValidYCoordinate = (yCoordinate: number, board: BoardType) : boolean => {
  if (typeof yCoordinate !== 'number' || isNaN(yCoordinate) || yCoordinate < 0 || yCoordinate >= board.yWidth) {
    return false;
  }

  return true;
};

const isValidDirection = (direction: DirectionType, possibleDirections: DirectionType[]) : boolean => {
  if (typeof direction !== 'string' || !possibleDirections.includes(direction)) {
    return false;
  }

  return true;
};

class Robot {
  private xPosition?: number;
  private yPosition?: number;
  private direction?: DirectionType;
  private board: BoardType;
  private possibleDirections : DirectionType[];
  private isRobotPlaced;

  constructor(board: BoardType, possibleDirections: DirectionType[]) {
    this.board = board;
    this.possibleDirections = possibleDirections;
    this.isRobotPlaced = isRobotPlaced(board, possibleDirections);
  }

  place = (x: number, y: number, facing: DirectionType) : void => {
    if (!isValidXCoordinate(x, this.board)) {
      throw new Error('X Coordinate is invalid');
    }
    if (!isValidYCoordinate(y, this.board)) { 
      throw new Error('Y Coordinate is invalid');
    }
    if (!isValidDirection(facing, this.possibleDirections)) {
      throw new Error('The given direction, the Robot is facing, is invalid');
    }

    this.xPosition = x;
    this.yPosition = y;
    this.direction = facing;
  };

  move = () : void => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    switch (this.direction) {
      case 'NORTH':
        if (typeof this.yPosition !== 'number' || !isValidYCoordinate(this.yPosition + 1, this.board)) {
          throw new Error('The robot can\'t move in that direction, or it falls down');
        }
        ++this.yPosition;
        break;
      case 'EAST':
        if (typeof this.xPosition !== 'number' || !isValidXCoordinate(this.xPosition + 1, this.board)) {
          throw new Error('The robot can\'t move in that direction, or it falls down');
        }
        ++this.xPosition;
        break;
      case 'SOUTH':
        if (typeof this.yPosition !== 'number' || !isValidYCoordinate(this.yPosition - 1, this.board)) {
          throw new Error('The robot can\'t move in that direction, or it falls down');
        }
        --this.yPosition;
        break;
      case 'WEST':
        if (typeof this.xPosition !== 'number' || !isValidXCoordinate(this.xPosition - 1, this.board)) {
          throw new Error('The robot can\'t move in that direction, or it falls down');
        }
        --this.xPosition;
        break;
    }
  };

  left = () : void => {
    this.direction = this.nextDirection('left');
  };

  right = () : void => {
    this.direction = this.nextDirection('right');
  };

  report = () : PositionType => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    return {
      xPosition: this.xPosition,
      yPosition: this.yPosition,
      direction: this.direction,
      toString: () => `${this.xPosition}, ${this.yPosition}, ${this.direction}`,
    };
  };

  private nextDirection = (turnDirection: 'left'|'right') : undefined|DirectionType => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    switch (this.direction) {
      case 'NORTH':
        return turnDirection === 'left' ? 'WEST' : 'EAST';
      case 'EAST':
        return turnDirection === 'left' ? 'NORTH' : 'SOUTH';
      case 'SOUTH':
        return turnDirection === 'left' ? 'EAST' : 'WEST';
      case 'WEST':
        return turnDirection === 'left' ? 'SOUTH' : 'NORTH';
    }
  };
}

export default Robot;
export {
  defaultBoard,
  possibleDirections,
  isRobotPlaced,
  isValidXCoordinate,
  isValidYCoordinate,
  isValidDirection,
  BoardType,
  PositionType,
  DirectionType,
};