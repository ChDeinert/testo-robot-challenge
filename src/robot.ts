type Board = { xWidth: number, yWidth: number };
type Direction = 'NORTH' | 'EAST' | 'SOUTH' | 'WEST';
type Position = { xPosition: undefined|number, yPosition: undefined|number, direction: undefined|Direction, toString?: () => string };

const defaultBoard : Board = {
  xWidth: 5,
  yWidth: 5,
};

const possibleDirections = [
  'NORTH', 'EAST', 'SOUTH', 'WEST'
];

const isRobotPlaced = (board : Board, possibleDirections: Direction[]) => (xCoordinate : undefined|number, yCoordinate : undefined|number, direction: undefined|Direction) : boolean => {
  if (!isValidXCoordinate(xCoordinate, board) || !isValidYCoordinate(yCoordinate, board) || !isValidDirection(direction, possibleDirections)) {
    return false;
  }

  return true;
};

const isValidXCoordinate = (xCoordinate : number, board: Board) : boolean => {
  if (typeof xCoordinate !== 'number' || xCoordinate < 0 || xCoordinate >= board.xWidth) {
    return false;
  }

  return true;
};

const isValidYCoordinate = (yCoordinate : number, board: Board) : boolean => {
  if (typeof yCoordinate !== 'number' || yCoordinate < 0 || yCoordinate >= board.yWidth) {
    return false;
  }

  return true;
};

const isValidDirection = (direction : Direction, possibleDirections: Direction[]) : boolean => {
  if (typeof direction !== 'string' || !possibleDirections.includes(direction)) {
    return false;
  }

  return true;
};

class Robot {
  private xPosition: undefined | number;
  private yPosition: undefined | number;
  private direction: undefined | Direction;
  private board: Board;
  private possibleDirections : Direction[];
  private isRobotPlaced;

  constructor(board: Board, possibleDirections : Direction[]) {
    this.board = board;
    this.possibleDirections = possibleDirections;
    this.isRobotPlaced = isRobotPlaced(board, possibleDirections);
  }

  place = (x: number, y: number, facing: Direction) : void => {
    if (!isValidXCoordinate(x, this.board)) {
      throw new Error('X Coordinate is invalid');
    }
    if (!isValidYCoordinate(y, this.board)) { 
      throw new Error('Y Coordinate is invalid');
    }
    if (!isValidDirection(facing, this.possibleDirections)) {
      throw new Error('The given direction, the Robot is facing, is invalid');
    }

    this.xPosition = x;
    this.yPosition = y;
    this.direction = facing;
  };

  move = () : void => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }
  };

  left = () : void => {
    this.direction = this.nextDirection('left');
  };

  right = () : void => {
    this.direction = this.nextDirection('right');
  };

  report = () : Position => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    return {
      xPosition: this.xPosition,
      yPosition: this.yPosition,
      direction: this.direction,
      toString: () => `${this.xPosition}, ${this.yPosition}, ${this.direction}`,
    };
  };

  private nextDirection = (turnDirection : 'left'|'right') : Direction => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    switch (this.direction) {
      case 'NORTH':
        return turnDirection === 'left' ? 'WEST' : 'EAST';
      case 'EAST':
        return turnDirection === 'left' ? 'NORTH' : 'SOUTH';
      case 'SOUTH':
        return turnDirection === 'left' ? 'EAST' : 'WEST';
      case 'WEST':
        return turnDirection === 'left' ? 'SOUTH' : 'NORTH';
    }
  };
};

export default Robot;
export { defaultBoard, possibleDirections, isRobotPlaced, isValidXCoordinate, isValidYCoordinate, isValidDirection };