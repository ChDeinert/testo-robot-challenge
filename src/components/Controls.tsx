import React, { useCallback } from 'react';
import { useForm, SubmitHandler } from "react-hook-form";
import { useRobotContext } from '../context/RobotContext';
import { DirectionType } from '../lib/robot';
import SVGButton from './SVGButton';
import ArrowLeft from '../assets/svg/duotone-undo-svgrepo-com.svg';
import ArrowUp from '../assets/svg/duotone-arrow-up-svgrepo-com.svg';
import ArrowRight from '../assets/svg/duotone-redo-svgrepo-com.svg';
import QuestionMark from '../assets/svg/duotone-question-small-svgrepo-com.svg';
import './Controls.css';

type FormValues = {
  xCoordinate: string;
  yCoordinate: string;
  direction: string;
};

const Controls : React.FC = () => {
  const { 
    settings: { board, possibleDirections }, 
    state: { initialized }, 
    controls
  } = useRobotContext();
  const { register, handleSubmit, formState: { errors } } = useForm();

  const onPlace : SubmitHandler<FormValues> = useCallback(({ xCoordinate, yCoordinate, direction } : FormValues) => {
    controls?.place(parseInt(xCoordinate), parseInt(yCoordinate), direction as DirectionType);
  }, [ controls ]);
  const onMoveClick = useCallback(() => {
    controls?.move();
  }, [ controls ]);
  const onLeftClick = useCallback(() => {
    controls?.turnLeft();
  }, [ controls ]);
  const onRightClick = useCallback(() => {
    controls?.turnRight();
  }, [ controls ]);
  const onReportPositionClick = useCallback(() => {
    const currentPositionReport = controls?.report();
    if (currentPositionReport) {
      alert(`The Robot's current position: ${currentPositionReport}`);
    }
  }, [ controls ]);

  return !initialized ? (<p>loading...</p>) : (
    <>
      <div className="controls__buttons" data-testid="robotcontrolbuttons">
        <SVGButton onClickHandler={onLeftClick} title="Turn left" data-testid="robotcontrol-button-turnleft">
          <ArrowLeft aria-hidden="true" />
        </SVGButton>
        <SVGButton onClickHandler={onMoveClick} title="Move Robot" data-testid="robotcontrol-button-move">
          <ArrowUp aria-hidden="true" />
        </SVGButton>
        <SVGButton onClickHandler={onRightClick} title="Turn right" data-testid="robotcontrol-button-turnright">
          <ArrowRight aria-hidden="true" />
        </SVGButton>
      </div>
      <div className="controls__buttons">
        <SVGButton onClickHandler={onReportPositionClick} data-testid="robotcontrol-button-report">
          <QuestionMark aria-hidden="true" /> Report current position
        </SVGButton>
      </div>
      <form onSubmit={handleSubmit(onPlace)} className="controls__form" data-testid="robotplaceform">
        <label htmlFor="xCoordinate" className="controls__form__group controls__form__label">
          X-Coordinate: 
          <input 
            type="number" 
            {...register('xCoordinate', { required: true, min: 0, max: board.xWidth - 1 })} 
            min={0} 
            max={board.xWidth - 1} 
            size={1} 
            className="controls__form__input" 
          />
          <span className="controls__form__group__warning">{errors.xCoordinate?.type === 'required' && 'X-Coordinate is required!'}</span>
        </label>
        <label htmlFor="yCoordinate" className="controls__form__group controls__form__label">
          Y-Coordinate: 
          <input 
            type="number" 
            {...register('yCoordinate', { required: true, min: 0, max: board.yWidth - 1 })} 
            min={0} 
            max={board.yWidth - 1} 
            size={1} 
            className="controls__form__input" 
          />
          <span className="controls__form__group__warning">{errors.yCoordinate?.type === 'required' && 'Y-Coordinate is required!'}</span>
        </label>
        <label htmlFor="direction" className="controls__form__group controls__form__label">
          Facing direction:
          <select {...register('direction', { required: true })} className="controls__form__input">
            {possibleDirections.map((direction, index) => (<option key={`direction_${index}`} value={direction}>{direction}</option>))}
          </select>
        </label>
        <input type="submit" value="Place Robot" className="controls__form__input" />
      </form>
    </>
  );
};

export default Controls;