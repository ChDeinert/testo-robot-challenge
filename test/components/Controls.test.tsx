import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import Controls from "../../src/components/Controls";
import RobotContext from "../../src/context/RobotContext";
import { defaultBoard, possibleDirections } from "../../src/lib/robot";

describe('<Controls />', () => {
  it('renders a loading indication if Robot is not initialized in context', async () => {
    render(<Controls />);
    const placeForm = screen.getByText('loading...');
    expect(placeForm).toMatchInlineSnapshot(`
      <p>
        loading...
      </p>
    `);
  });

  it('renders the place form if Robot is initialized in context', () => {
    const testContext = {
      settings: {
        board: defaultBoard,
        possibleDirections: possibleDirections,
      },
      state: {
        initialized: true,
        isPlaced: false,
        position: {
          xPosition: 0,
          yPosition: 0,
          direction: possibleDirections[0],
        },
      },
    };
    render(
      <RobotContext.Provider value={testContext}>
        <Controls />
      </RobotContext.Provider>
    );
    const placeForm = screen.getByTestId('robotplaceform');
    expect(placeForm).toMatchSnapshot();
  });

  it('renders the robot control if Robot is initialized in context', () => {
    const testContext = {
      settings: {
        board: defaultBoard,
        possibleDirections: possibleDirections,
      },
      state: {
        initialized: true,
        isPlaced: false,
        position: {
          xPosition: 0,
          yPosition: 0,
          direction: possibleDirections[0],
        },
      },
    };
    render(
      <RobotContext.Provider value={testContext}>
        <Controls />
      </RobotContext.Provider>
    );
    const placeForm = screen.getByTestId('robotcontrolbuttons');
    expect(placeForm).toMatchSnapshot();
  });
});
