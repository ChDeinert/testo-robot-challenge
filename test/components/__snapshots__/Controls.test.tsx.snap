// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Controls /> renders the place form if Robot is initialized in context 1`] = `
<form
  class="controls__form"
  data-testid="robotplaceform"
>
  <label
    class="controls__form__group controls__form__label"
    for="xCoordinate"
  >
    X-Coordinate:
    <input
      class="controls__form__input"
      max="4"
      min="0"
      name="xCoordinate"
      size="1"
      type="number"
    />
    <span
      class="controls__form__group__warning"
    />
  </label>
  <label
    class="controls__form__group controls__form__label"
    for="yCoordinate"
  >
    Y-Coordinate:
    <input
      class="controls__form__input"
      max="4"
      min="0"
      name="yCoordinate"
      size="1"
      type="number"
    />
    <span
      class="controls__form__group__warning"
    />
  </label>
  <label
    class="controls__form__group controls__form__label"
    for="direction"
  >
    Facing direction:
    <select
      class="controls__form__input"
      name="direction"
    >
      <option
        value="NORTH"
      >
        NORTH
      </option>
      <option
        value="EAST"
      >
        EAST
      </option>
      <option
        value="SOUTH"
      >
        SOUTH
      </option>
      <option
        value="WEST"
      >
        WEST
      </option>
    </select>
  </label>
  <input
    class="controls__form__input"
    type="submit"
    value="Place Robot"
  />
</form>
`;

exports[`<Controls /> renders the robot control if Robot is initialized in context 1`] = `
<div
  class="controls__buttons"
  data-testid="robotcontrolbuttons"
>
  <button
    class="svgbutton"
    data-testid="robotcontrol-button-turnleft"
  >
    <duotone-undo-svgrepo-com.svg
      aria-hidden="true"
    />
    <span
      class="svgbutton--visually-hidden"
    >
      Turn left
    </span>
  </button>
  <button
    class="svgbutton"
    data-testid="robotcontrol-button-move"
  >
    <duotone-arrow-up-svgrepo-com.svg
      aria-hidden="true"
    />
    <span
      class="svgbutton--visually-hidden"
    >
      Move Robot
    </span>
  </button>
  <button
    class="svgbutton"
    data-testid="robotcontrol-button-turnright"
  >
    <duotone-redo-svgrepo-com.svg
      aria-hidden="true"
    />
    <span
      class="svgbutton--visually-hidden"
    >
      Turn right
    </span>
  </button>
</div>
`;
